feat(kiali): traffic graph for cards

leandroberetta opened the pull request:
Kiali graph in the entity view (just the card):

![image](https://github.com/janus-idp/backstage-plugins/assets/1286393/b6a52ecd-20db-40a1-b4a5-f82d9830e6a5)

Entity card for dynamic use added:

![image](https://github.com/janus-idp/backstage-plugins/assets/1286393/a26b0313-4a6e-47d4-a3c2-6c1f0b052514)

ScriptingShrimp commented:
Hello @leandroberetta,
since main branch have the playwright tests already fixed, can you please add some regression before merging this PR? 
Thank you :pray:

leandroberetta commented:
> Hello @leandroberetta, since main branch have the playwright tests already fixed, can you please add some regression before merging this PR? Thank you 🙏

We decided to merge this as this has been open for a long time. I'm working to add a test in a follow up PR.

ScriptingShrimp commented:
> > Hello @leandroberetta, since main branch have the playwright tests already fixed, can you please add some regression before merging this PR? Thank you 🙏
> 
> We decided to merge this as this has been open for a long time. I'm working to add a test in a follow up PR.

Thank you :+1:

josunect reviewed (commented):
Thanks, @leandroberetta ! It is looking good: 

![image](https://github.com/janus-idp/backstage-plugins/assets/49480155/8bebe0a4-64a0-4561-a5af-2ce89827ddcd)

![image](https://github.com/janus-idp/backstage-plugins/assets/49480155/a61d9a20-149b-4373-be60-580838a61835)

But, I've found a confusing behavior: the graph is refreshed ~3 times, in both pages the Kiali page and in the cards. Also, in the cards view, when the user selects a different resource the graph is also refreshed, but I don't think it should be, right? 

[Screencast from 04-07-24 12:08:50.webm](https://github.com/janus-idp/backstage-plugins/assets/49480155/50435817-1242-4a47-b6f8-e3ed57c82ae6)

aljesusg commented:
This need to be rebased (Error about kiali version). I am going to install that verswion to test it

aljesusg commented:
# Kiali standalone

looks good. I would move the trafic graph tab to the second postiipon like kiali
![image](https://github.com/user-attachments/assets/88bc701d-78d7-4aff-85fc-f77335906531)

# Entity view
As @josunect  said not working with namespace selector.
Fix the height to be aligned please.
+ It would be gret to have some kind of zoom button but I imagine that is going to take more job to do that right? We can do it later
![image](https://github.com/user-attachments/assets/bff38b8a-9a90-4a84-be42-062bee82aa68)

Graph card not work for me. If we are going to add  this in this PR I would add some tests due this will be supported.

I'll review the code later if I have time before my PTO

leandroberetta commented:
> This need to be rebased (Error about kiali version). I am going to install that verswion to test it

I'm using v1.86 to test this, isn't ok?.

leandroberetta commented:
> # Kiali standalone
> looks good. I would move the trafic graph tab to the second postiipon like kiali ![image](https://private-user-images.githubusercontent.com/3019213/348243121-88bc701d-78d7-4aff-85fc-f77335906531.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3ODc1NDIsIm5iZiI6MTcyMDc4NzI0MiwicGF0aCI6Ii8zMDE5MjEzLzM0ODI0MzEyMS04OGJjNzAxZC03OGQ3LTRhZmYtODVmYy1mNzczMzU5MDY1MzEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTJUMTIyNzIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGE5NmYwNDc3MmUxMmU2Zjc2MjY4OWM4ZDZjOGI0NjQ2OWNiMGVmZTk2ZDUwMDA1YzEzOTJkOGQwZmZmMzM1OSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.W48mPQBebqwB1-oZgzSmmLQIDdK4kZWXw0lo09Eg8_s)
> 
> # Entity view
> As @josunect said not working with namespace selector. Fix the height to be aligned please.
> 
> * It would be gret to have some kind of zoom button but I imagine that is going to take more job to do that right? We can do it later
>   ![image](https://private-user-images.githubusercontent.com/3019213/348243638-bff38b8a-9a90-4a84-be42-062bee82aa68.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3ODc1NDIsIm5iZiI6MTcyMDc4NzI0MiwicGF0aCI6Ii8zMDE5MjEzLzM0ODI0MzYzOC1iZmYzOGI4YS05YTkwLTRhODQtYmU0Mi0wNjJiZWU4MmFhNjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTJUMTIyNzIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDQ1Y2Y4ZmJhMGM5NDNhNmYzMzMyOTVjM2JiOTA3YjgyMTljODY2NzhmODBmODA2MWNmZjNjNzEwNDEzYTA1MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.C7h6UCxxyh6uEFJRo13IHewMzbsHrbOM7mb-8mVjOlA)
> 
> * Graph card not work for me. If we are going to add  this in this PR I would add some tests due this will be supported.
> * On the other hand we should be able to adap theme background to light/dark like we do with other cards an kiali standalone
> 
> # Dev environment
> Pelase and the graph card like we did with the resources card ![image](https://private-user-images.githubusercontent.com/3019213/348248867-fca90856-7fd0-414d-b438-7747d8c08cba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3ODc1NDIsIm5iZiI6MTcyMDc4NzI0MiwicGF0aCI6Ii8zMDE5MjEzLzM0ODI0ODg2Ny1mY2E5MDg1Ni03ZmQwLTQxNGQtYjQzOC03NzQ3ZDhjMDhjYmEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MTJUMTIyNzIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmExYWIwYjc0YTliMGU4MjlhNTQ3OTA0Nzk5YWFkNWY5NGMzNzA1ZjIxOGNjYmY3MTgwYzQxNDlhMjA5ZTJmYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.RWknPzL5ZFhhVNxpoxLAyjxSHjlEV7QlbJbt6abU7IA)
> 
> I'll review the code later if I have time before my PTO

Thanks for review. I will address suggestions.

leandroberetta commented:
I addressed, I think, everything needed:

- Added minimal testing (more will be added on follow up PR)
- Added dark mode support
- Added a minimal control bar (zoom in/out, etc)
- Fix the issue with the namespace selector on the entity view

![image](https://github.com/user-attachments/assets/49db275d-c0c0-4616-bef3-1c4029a661a3)

josunect commented on plugins/kiali/src/pages/TrafficGraph/TrafficGraphCard.tsx:
```diff
@@ -0,0 +1,21 @@
+  const { entity } = useEntity();
+
+  return (
+    <Card style={{ marginRight: 20, marginBottom: 20, height: 425.59 }}>
```
Why is this margin right? I think it is not looking good for the Kiali tab:

![image](https://github.com/user-attachments/assets/fb8967e1-35b6-43b2-8416-8c6807a4be7a)

josunect commented:
Thanks @leandroberetta , some comments were addressed: 

- The dark mode was fixed
![image](https://github.com/user-attachments/assets/cb0f2956-7ee7-4595-9614-6c6d215e0383)

- Graph card added
![image](https://github.com/user-attachments/assets/228c4993-b981-4726-9fdd-e501bd6d0e69)

- zoom buttons added
- namespace selector is working now. 
- Initial test added


I have some comments: 

- The graph is not loaded when you refresh the page (For the resources card): 
![image](https://github.com/user-attachments/assets/d93ab1df-38ac-4a69-b492-de5affbecb1f)

- One comment from @aljesusg missing: Moving the Kiali graph tab to the second position. 
- As they were some failures in the refresh, I would add at least one more test to verify there are some nodes in the graph, for the graph card.

leandroberetta commented:
> Thanks @leandroberetta , some comments were addressed:
> 
> * The dark mode was fixed
>   ![image](https://private-user-images.githubusercontent.com/49480155/349460849-cb0f2956-7ee7-4595-9614-6c6d215e0383.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNjExNTksIm5iZiI6MTcyMTM2MDg1OSwicGF0aCI6Ii80OTQ4MDE1NS8zNDk0NjA4NDktY2IwZjI5NTYtN2VlNy00NTk1LTk2MTQtNmM2ZDIxNWUwMzgzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDAzNDczOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0MDNkOGMyNzA1NTY1N2Q5MWJkMDJkMWMzODM5NTI5MjQ2Mjk0MDViZDE5MTQ0MWJkZmZkNDA4NzYzN2M1ZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vz-lSNSqnZtx77mCANNOqLXfirlRtaQD4mtgBSQzbZ0)
> * Graph card added
>   ![image](https://private-user-images.githubusercontent.com/49480155/349461125-228c4993-b981-4726-9fdd-e501bd6d0e69.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNjExNTksIm5iZiI6MTcyMTM2MDg1OSwicGF0aCI6Ii80OTQ4MDE1NS8zNDk0NjExMjUtMjI4YzQ5OTMtYjk4MS00NzI2LTlmZGQtZTUwMWJkNmQwZTY5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDAzNDczOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjZTU2YzQ1YjBhNzdkYTcxNGJiZmFlMjA5NzVlODA3NDU4YjM5MTRmODVmYmQwMzhmM2Y4MGYzYzc0ZDJhNGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bS7S3hHt7FxopKZl3uD8XERdNDW_SehHeYw74XTxv0I)
> * zoom buttons added
> * namespace selector is working now.
> * Initial test added
> 
> I have some comments:
> 
> * The graph is not loaded when you refresh the page (For the resources card):
>   ![image](https://private-user-images.githubusercontent.com/49480155/349462548-d93ab1df-38ac-4a69-b492-de5affbecb1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzNjExNTksIm5iZiI6MTcyMTM2MDg1OSwicGF0aCI6Ii80OTQ4MDE1NS8zNDk0NjI1NDgtZDkzYWIxZGYtMzhhYy00YTY5LWI0OTItZGU1YWZmYmVjYjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDAzNDczOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY1ZDA4MWJhNzc5OTg5NDM1ZDYwYWE4MGZlYmJiYmNjOWYzOTM0ZGUyNmI5YmM5OGQwZWFlNmZmOTQzOGEzMjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.v48z2rx_aErzCR5rzEZ3ReWH68Pe80gtQRJpZBy0RG8)
> * One comment from @aljesusg missing: Moving the Kiali graph tab to the second position.
> * As they were some failures in the refresh, I would add at least one more test to verify there are some nodes in the graph, for the graph card.

Hi @josunect thanks for reviewing. I addressed the suggestions.

I've been trying to get the test pass but it seems Github is not working properly, task are being killed, but I know from several passes that lint is ok and testing too. 
I'll keep trying but you can review the code.

josunect reviewed (approved):
LGTM!

All comments were addressed.

I would open a follow up PR to update the documentation. 
I would add in the dev environement the travels data, to be coherence with the other examples.

aljesusg commented:
Please @leandroberetta add this to the readme docs with some screenshots . Good Job !
